src/driver.ml

```diff
@@ -556,92 +563,124 @@ let print_passes () =
       Printf.printf "<builtin:check-unused-extensions>\n")
 
 (*$*)
+
+exception WrapStructureAndLintErrors of (structure * attribute list) * exn
+
+exception WrapStructure of structure * exn
+
 let map_structure_gen st ~tool_name ~hook ~expect_mismatch_handler ~input_name =
   Cookies.acknowledge_cookies T;
   if !perform_checks then (
     Attribute.reset_checks ();
     Attribute.collect#structure st);
+  let lint_and_check lint_errors st =
+    let st =
+      match lint_errors with
+      | [] -> st
+      | _ ->
+          List.map lint_errors
+            ~f:(fun ({ attr_name = { loc; _ }; _ } as attr) ->
+              Ast_builder.Default.pstr_attribute ~loc attr)
+          @ st
+    in
+    Cookies.call_post_handlers T;
+    if !perform_checks then (
+      (* TODO: these two passes could be merged, we now have more passes for
+         checks than for actual rewriting. *)
+      Attribute.check_unused#structure st;
+      if !perform_checks_on_extensions then Extension.check_unused#structure st;
+      Attribute.check_all_seen ();
+      if !perform_locations_check then
+        let open Location_check in
+        ignore
+          ((enforce_invariants !loc_fname)#structure st
+             Non_intersecting_ranges.empty
+            : Non_intersecting_ranges.t));
```

Comment:
You are right :dart:.

I also agree that we should embed the generated errors in the AST. I can do that in this PR as well.

Reply:
In fact, I think it might be better to do it in another PR, focused on replacing the `raise_errorf` calls by `error_extensionf`, when PR #316 has been merged. (I count 34 calls of `raise_errorf`).